### ISSUE TITLE
Reading of uninitialized local variables is broken

### ISSUE THREAD

**Issue by smarr** [SOM-st/TruffleSOM#6]
Reading of uninitialized variables in methods is likely broken. At least it is in [SOMns](https://github.com/smarr/SOMns/issues/40).

The following test fails with an assertion, because the read evaluates to `null`:

```Smalltalk
class Test usingPlatform: platform = Value ()(
  protected foo = (
    | a |
    a println.
    a := 1.
    a println.
  )

  public main: args = (
    foo.
    foo.
    ^ 0
  )
)
```

A fix could be:

```patch
diff --git a/src/som/compiler/MethodBuilder.java b/src/som/compiler/MethodBuilder.java
index 1598a2b..20e8757 100644
--- a/src/som/compiler/MethodBuilder.java
+++ b/src/som/compiler/MethodBuilder.java
@@ -101,7 +102,8 @@ public final class MethodBuilder {
     MethodScope outer = (outerBuilder != null)
         ? outerBuilder.getCurrentMethodScope()
         : null;
-    this.currentScope   = new MethodScope(new FrameDescriptor(), outer, clsScope);
+    assert Nil.nilObject != null;
+    this.currentScope   = new MethodScope(new FrameDescriptor(Nil.nilObject), outer, clsScope);
 
     accessesVariablesOfOuterScope = false;
     throwsNonLocalReturn          = false;
```

/cc @charig

**Comment by charig** [comment]
Stefan, good catch, same happens in TruffleMate and I guess in TruffleSOM. But, have you tried the proposed patch? Anyway, my impression is that each LocalVarNode* should have its own Slot (clone the slot at creation time) while they are sharing it right now.

**Comment by smarr** [comment]
The patch is for SOMns, it fixed the example above. I haven't tested it beyond that.

I am not sure exactly what you are referring to with the LocalVarNode* and having their own slot.
The slot represents the memory for a local variable. The LocalVarNode*s are there to access them. Cloning the slots does not make sense to me. What would that achieve? That would be similar to duplicating a field in an object for each place it is accessed, no?

The frame is actually tracking whether the slot is set. It has an array of tags. That could potentially also be used to mark a variable as unwritten again (for issue #5).

**Comment by charig** [comment]
Yes. What I said was complete nonsense. It would even be incorrect. What I was trying to think is something like this:  ` 
protected foo = (
    | a |
    a := nil.
    a println.
    a := 1.
    a println.
  )
`
The second LocalVarNode specializes to object because the FrameSlotKind is already set to object by nil. While debugging your example, I was thinking in a workaround to try to optimize this case but clearly duplicating FrameSlots is not the solution, sorry for the noise.

**Comment by charig** [comment]
If you want to update TruffleSOM it is quite straightforward but just in case i attach a: [commit](https://github.com/charig/TruffleMATE/commit/f053cb3b1702bc7d7f38abd33781315d821a9f5a). and its regression [test](https://github.com/charig/SOM/commit/ba8f02e3d5edb126168a6ba7813febd57430c8c0)
